### PR TITLE
Add signature verifier to signingalgorithm.go

### PR DESCRIPTION
**Pull request by irori**
This also moves signer_test.go to signingalgorithm_test.go, as it only uses `signingalgorithm` package.

### CODE REVIEW

File: go/signedexchange/internal/signingalgorithm/signingalgorithm.go [hunk 1]
```diff
@@ -20,11 +23,11 @@ type ecdsaSigningAlgorithm struct {
 	rand    io.Reader
 }
 
-func (e *ecdsaSigningAlgorithm) Sign(m []byte) ([]byte, error) {
-	type ecdsaSigValue struct {
-		R, S *big.Int
-	}
+type ecdsaSigValue struct {
```
Comment (nyaxt) [inline go/signedexchange/internal/signingalgorithm/signingalgorithm.go]: Optional: add RFC 3279 ref?

File: go/signedexchange/internal/signingalgorithm/signingalgorithm.go [hunk 2]
```diff
@@ -46,5 +49,44 @@ func SigningAlgorithmForPrivateKey(pk crypto.PrivateKey, rand io.Reader) (Signin
 			return nil, fmt.Errorf("signedexchange: unknown ECDSA curve: %s", name)
 		}
 	}
-	return nil, fmt.Errorf("signedexchange: unknown public key type: %T", pk)
+	return nil, fmt.Errorf("signedexchange: unknown private key type: %T", pk)
```
Comment (nyaxt) [inline go/signedexchange/internal/signingalgorithm/signingalgorithm.go]: Can be another PR: prefix all error message with "signingalgorithm: "?

**Comment by irori** [comment]
(This is part of #319)

File: go/signedexchange/internal/signingalgorithm/signingalgorithm.go [hunk 3]
```diff
@@ -43,8 +47,47 @@ func SigningAlgorithmForPrivateKey(pk crypto.PrivateKey, rand io.Reader) (Signin
...
+	if len(rest) > 0 {
+		return false, errors.New("signingalgorithm: extra data at the signature end")
+	}
+
+	hash := e.hash.New()
+	hash.Write(msg)
+	return ecdsa.Verify(e.pubKey, hash.Sum(nil), v.R, v.S), nil
+}
+
+func VerifierForPublicKey(k crypto.PublicKey) (Verifier, error) {
```
Comment (hajimehoshi) [inline go/signedexchange/internal/signingalgorithm/signingalgorithm.go]: Do we plan to add another verifier other than `ecdsaVerifier`? If not, I think exporting `ecdsaVerifier` and returning it instead of returning an interface value should be fine.

Comment (irori) [reply go/signedexchange/internal/signingalgorithm/signingalgorithm.go]: We may support ed25519 (and/or other non-legacy, non-RSA signing algorithms defined by TLS 1.3 or later) in the future.

I'd like to use this interface in the first place, to minimize caller-side changes when we support them.

Comment (hajimehoshi) [reply go/signedexchange/internal/signingalgorithm/signingalgorithm.go]: Makes sense.

File: go/signedexchange/internal/signingalgorithm/signingalgorithm_test.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+)
+
+func TestSignVerify_ECDSA_P256_SHA256(t *testing.T) {
+	pk, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		t.Fatalf("Failed to generate ecdsa private key: %v", err)
+		return
+	}
+
+	alg, err := SigningAlgorithmForPrivateKey(pk, rand.Reader)
```
[the remark this answers is not in the document]
Comment (hajimehoshi) [reply go/signedexchange/internal/signingalgorithm/signingalgorithm_test.go]: Oh, I missed that, sorry
